### ISSUE TITLE
Double hermgauss

### ISSUE THREAD

**Issue by Bonnevie** [GPflow/GPflow#306]
There appears to be two identical functions called hermgauss - one in likelihoods.py and one in kernels.py. I assume this is not intentional. Other files like ekernels.py import from kernels.py.

**Comment by jameshensman** [comment]
Thanks for the report, @Bonnevie . You're right of course. 

The source of the confusion is that @markvdw implemented the multi-dimensional quadrature code for ekernels.py, and also reused the hermgauss function. It would be nice to tidy the code a little to unify these. 

Mark, since you know your way around these bits of code better than most, I've assigned you. I don't think this is urgent: perhaps we could flag it as a 'junior job' that people who wanted to start hacking on the codebase could handle, with your supervision?

**Comment by markvdw** [comment]
Ah yes, I remember doing this and knowing it wasn't particularly neat. Perhaps the best solution would be to have a separate `quadrature.py`?

I agree it's not super high priority, and if you know anyone who wants to submit a patch, that would be great. Otherwise I can do it.

**Comment by Bonnevie** [comment]
A separate quadrature.py would make a lot of sense. 

Related issue: I was just in the process of figuring out how mvhermgauss works and from an object-oriented perspective it seems a bit off. Since it's made to calculate N simultaneous quadratures it collects N separate grids in a single array. This is of course sensible since it allows a function to be applied to all evaluation points simultaneously, but it also means there are additional post-processing steps that need to be performed after every call to the function. Furthermore, I wouldn't say it's immediately apparent how you construct N separate quadratures from arrays with shape (herm_points^(input_dim) * N ) x input_dim in an unambigious manner without consulting the source and/or prior applications of mvhermgauss.

Copying a pattern from kernels.py, I think it would make sense to have a function of the form (maybe with a few more bells and whistles):
```
def mvquad(f, Xmu, Xcov, H, D):
'''compute N multivariate Gaussian expectations using Gauss-Hermite quadrature.
f - function object. Function to integrate.
Xmu - NxD tensor. Means of each of the N gaussians to integrate over.
Xcov - NxDxD tensor. Covariance matrices of each of the N gaussians to integrate over.
H - Integer. Number of points used for the quadrature along each dimension.
D - Integer. Number of input dimensions. 
'''
    X, wn = mvhermgauss(Xmu, Xcov, H, D)  # (H**DxNxD, H**D)
    fX = tf.reshape(f(X), (H ** D, tf.shape(Xmu)[0]))
    return tf.reduce_sum(fX * wn[:, None], 0)

```

**Comment by markvdw** [comment]
So, I'm not completely happy with `mvhermgauss()` yet, and I agree with your objection. I would add the following "small grievances":
- Does not take numpy array arguments. This would be useful for debugging.
- I think it has a different convention to `hermgauss()`. `mvhermgauss()` assumes you're integrating against a Gaussian distribution. `hermgauss()` does regular Hermite quadrature, so extra rescaling terms are needed. I'd argue that for our purposes, a convention of integrating against a Gaussian distribution is probably most useful.
- The shape issue you mention.

One solution:
- Have `mvhermgauss()` calculate the grid and weights for a _standard_ Gaussian only. This removes the cholesky inside the function, and makes the interface simpler in exactly the way you want.
- Add a `mvquad()` function, as you suggested, to actually _perform_ the quadrature on a function. `f`  can be passed in as a `lambda`.

The remaining issue is that of which convention to use. @jameshensman, what do you think?

@Bonnevie: Since you've pretty much already come up with the solution, do you want to do the pull request?

**Comment by Bonnevie** [comment]
- Could the autoflow functionality be used to make a Numpy variant?
- You're right that hermgauss() and mvhermgauss() follow quite different conventions despite the misleadingly similar naming pattern; hermgauss() is in fact called internally in mvhermgauss prior to the application of the transformations. 
- I think the shape issue is mostly an issue because of a lack of abstraction. The steps implemented by mvhermgauss are necessary for making the quadrature and constitute a self-contained computational task. But actually performing the quadrature is left to the programmer. Adding a mvquad() solves the issue in my opinion.
 
- I don't think that rewriting mvhermgauss() to only handle a single standard gaussian integral is particularly helpful since this is functionality that mvhermgauss() already implements, with none of the ambiguity incurred when performing simultaneous quadratures (if you only pass it one mean and covariance, you know that you need to sum over all outputted evaluation points to calculate the quadrature).
- I think the second solution is pretty sensible (no surprise), unless there are some use cases I haven't noticed - I don't know whether the main authors want to weigh in on that.
 
I could set up a quadrature.py, sure. Making the switch to the new functions across all the old files seems a bit more daunting as it involves altering a significant number of core functions. But I guess judicious use of the testing suites makes for a sufficient safeguard.

**Comment by markvdw** [comment]
I do think that rewriting `mvhermgauss()` to only take a single standard Gaussian (or unnormalised squared exponential) is a good idea, as this would simultaneously pull its functionality in line with `hermgauss()`, and it would be pure numpy.

Then we could add an extra function that would actually perform the quadrature, adding the abstraction that you suggest.

So, to summarise, the way I see it:
- Making `mvhermgauss()` create a weighting for a standard normal or exp(x²). Just numpy code. We could change `hermgauss()` to have the same convention or vice versa.
- Make a `mvquad(f, Xmu, Xcov, H, D)` function, to actually perform the quadrature for multiple Gaussian distributions. This will take a TensorFlow function for f.

This covers the use cases that I imagine:
- Bringing the meaning of `mvhermgauss()` in line with `hermgauss()`
- Doing some quadrature by hand with a numpy function for debugging / testing, using `mvnhermgauss()`.
- Doing the quadrature within GPflow using TensorFlow using `mvquad()`.

I don't think an AutoFlow variant for `mvquad()` would work. It would need to take a function `f` that only did _numpy_ operations. However, its input inside the function would be a TensorFlow variable.

It's probably not too bad to do the refactor. I usually just do a 'find all reference' search in PyCharm or your favourite IDE.

**Comment by Bonnevie** [comment]
Just for reference, the functions are as follows:

```
def hermgauss(n):
    x, w = np.polynomial.hermite.hermgauss(n)
    x, w = x.astype(np_float_type), w.astype(np_float_type)
    return x, w


def mvhermgauss(means, covs, H, D):
    """
    Return the evaluation locations, and weights for several multivariate Hermite-Gauss quadrature runs.
    :param means: NxD
    :param covs: NxDxD
    :param H: Number of Gauss-Hermite evaluation points.
    :param D: Number of input dimensions. Needs to be known at call-time.
    :return: eval_locations (H**D*NxD), weights (H**D)
    """
    N = tf.shape(means)[0]
    gh_x, gh_w = hermgauss(H)
    xn = np.array(list(itertools.product(*(gh_x,) * D)))  # H**DxD
    wn = np.prod(np.array(list(itertools.product(*(gh_w,) * D))), 1)  # H**D
    cholXcov = tf.cholesky(covs)  # NxDxD
    X = 2.0 ** 0.5 * tf.batch_matmul(cholXcov, tf.tile(xn[None, :, :], (N, 1, 1)),
                                     adj_y=True) + tf.expand_dims(means, 2)  # NxDxH**D
    Xr = tf.reshape(tf.transpose(X, [2, 0, 1]), (-1, D))  # (H**D*N)xD
    return Xr, wn * np.pi ** (-D * 0.5)
```

Making `mvhermgauss()` a Numpy function will align the functions less, not more, since `hermgauss()` is a Tensorflow wrapper around a Numpy function. 

I think the simplest solution to the naming issue would be to just change the name of `mvhermgauss()` to `mvquadgrid()` or something like that. If you want a refactoring, this first part:
```
    N = tf.shape(means)[0]
    gh_x, gh_w = hermgauss(H)
    xn = np.array(list(itertools.product(*(gh_x,) * D)))  # H**DxD
    wn = np.prod(np.array(list(itertools.product(*(gh_w,) * D))), 1)  # H**D
```
is only Numpy and performs a single closed computational task. The transformations could then be applied to the output using the remaining lines in a separate function.

Would it suffice to do the above refactoring and have a Numpy version of the TF transform-and-take-quadrature functions?

**Comment by markvdw** [comment]
`hermgauss()` is a fully numpy function. You're right that it's a wrapper in GPflow, which simply ensures that it returns something of the correct data type, but it's all numpy code. When I'm debugging, I sometimes like to write a numpy function that I want to do quadrature on, and it would be nice if there was a multi variate function to generate the quadrature grid for numpy as well. That's why, I reckon we should do the following changes:
- Change `mvhermgauss` to `mvhermgauss(n, D)`, where `n` is the number of quadrature points in each dimension. It should return a list of n^D points and weights, in numpy arrays. This brings it in line with `hermgauss`, which also returns numpy arrays.
- Change `mvhermgauss` to calculate unnormalised Gauss-Hermite weights, instead of normalised for a given Gaussian. This brings the convention in line with `hermgauss()` as well. I think it's best to not add any normalisation constant, since it depends on the covariance of the Gaussian, which the function doesn't yet have access to.

In terms of abstraction, these two functions only create the GH grids, and are fairly low-level. However, doing it this way, where we don't pass in any covariance information, gets rid of the shape problem, since these functions only ever do the quadrature for `exp(-x^T x)`. For GPflow, we want to multiple quadratures of the same function, with respect to normalised Gaussians with different means and covariances. For this we could add the `mvnquad` (Multi Variate Normal Quadrature) function, as you suggest.

So to summarise:
- `hermgauss()` and `mvhermgauss()` create Gauss-Hermite quadrature grids and are fully numpy.
- `mvnquad(f, means, covs, H, D)` actually performs the quadrature on `f` for each Gaussian specified in `means` and `covs`.

I don't think there's a need for a full function that does numpy quadrature. This would only be used for debugging purposes, and I'm happy as long as I can get the quadrature grid :) Does this sound in line with what you were thinking?

**Comment by Bonnevie** [comment]
you're of course right, I was working from some flawed memory that the second line did a conversion to TF... 

Just to make sure, you are advocating that `mvhermgauss()` returns a *single* multi-variate grid, right? 
This is different from the original function which returns N grid copies collected in a single array, but matches the small extracted block from my previous comment. 
Sounds sensible, I'll give it a go.

**Comment by markvdw** [comment]
Yes, I think that `mvhermgauss()` should return a single grid. I think this is the most sensible, as it only ever computes the grid for `exp(-x^T x)`, which is fixed.

I think the transformation of grids and weights is best grouped together with the quadrature itself. The transformation relies on all the same variables as the full quadrature, and we never need them separately.

**Comment by Bonnevie** [comment]
Okay, I have done the refactorization and corrected the existing code* so that all tests return OK. I haven't implemented any further tests as of yet since I take it quadrature is already being tested implicitly in the test suites. 

It turned out to be a bit more difficult than expected as many of the integrals were actually over D_out-dimensional outputs. But it worked out pretty nicely and I think overall it's an improvement. Maybe the code can be a tweaked a little more to make it even nicer.

*I did not fix ekernels yet, but I moved a copy of the old quad functions in there so that it still works.

**Comment by Bonnevie** [comment]
I don't think refactoring `ekernels.py` will be as simple as we've made it out to be. Since it calculates covariance, the function depends on the mean and the covariance matrix which we have so far assumed is only used to compute the evaluation points.
Can we make a dedicated quadrature function for covariances?

In the original code, it appears that they are deducting the means prior to the covariance as noted above, but then they add in the outer product of the means just before returning - shouldn't it be either or? Or is this somehow a more appropriate quadrature estimator?

**Comment by Bonnevie** [comment]
This issue should probably be closed.

**Comment by alexggmatthews** [comment]
@Bonnevie thanks very much
